Add mocking of commands - closes #2 and #17

mh182 opened the pull request:
Add an optional parameter to `mock_create [command]` to mock commands.

Adds following new functions:
- `mock_chroot` to create a deterministic system setup. It does not use `chown` - if someone comes up with a better name, suggestions are welcome.
- `path_override` and `path_rm` to manipulate `PATH`
- `mock_teardown` for cleanup mock objects

I updated the documentation with some examples how those functions can be used.

mh182 committed:
Double quote to prevent globbing and word splitting

mh182 committed:
Instead of 'echo $(cmd)', just use 'cmd'

mh182 committed:
Use 'grep -c' instead of 'grep|wc -l'

mh182 committed:
Fix BW01 warnings for 'Command not found'

mh182 committed:
mock_create: add optional argument to mock commands

mh182 committed:
mock_create: make index of created mock objects replicable

Remove local dependent code so that the index for creating new mocks is
deterministic (0, 5, 10, ...).

mh182 committed:
Add path_override to facilitate mocking system commands

mh182 committed:
Add path_rm to facilitate PATH modfications

mh182 committed:
Add mock_chroot for tests with pristine environments

mh182 committed:
Improve documentation how to provide a deterministic test setup

mh182 committed:
Add mock_teardown for easy cleanup of mock objects

jesperronn commented on test/mock_chroot.bats:
```diff
@@ -0,0 +1,115 @@
+  [[ -x "${output}/unlink" ]]
+  [[ -x "${output}/wc" ]]
+  [[ -x "${output}/which" ]]
+  [[ -x "${output}/xargs" ]]
```
I guess this list is specific to your machine, so perhaps a reduced set of commands would be enough to provide a decent test, so that this will pass even if run inside a plain Alpine docker image.


suggestion: `date`, `ln`, `ls`, `sh`, `rm`

mh182 replied:
Indeed, I just picked a set of commonly used commands and added those to the mocked chroot. The point of `mock_chroot` is to provide a default minimal system.

I don’t understand if you would like that I adapted the test, so it would pass on systems with a smaller footprint, or do you want a smaller minimal system?

jesperronn commented on src/bats-mock.bash:
```diff
@@ -57,7 +69,45 @@ fi
+# Globals:
+#   BATS_TMPDIR
+mock_teardown() {
+    rm -rf "${BATS_TMPDIR}"/bats-mock.$$.*
```
I could not find a unit test for `mock_teardown`. I think it would be really easy to provide one

mh182 replied:
Good catch, I will do that.

jesperronn reviewed (approved):
I took the liberty of reviewing the basic code changes of this pull-request, although I am in no way affilliated with [grayhemp](https://github.com/grayhemp)/[bats-mock](https://github.com/grayhemp/bats-mock) project.

Would be fine if @grayhemp or another maintainer could review. 

I might be misunderstanding, but for me it seems as I have to use this changeset to make mocks available inside each file under test. 

(altough there may be easier workarounds, which I have not yet discovered.)

I like the approach that mock_create injects a symlink to the mock into the path loaded for the file under test.

mh182 committed:
mock_chroot: make test case less pedantic for small systems

Reduce the number of mocked commands to a minimum. Smaller systems like
Alpine may not have all commands installed wich make the test case fail.

mh182 committed:
Fix test cases which break when executed with bats container image

The container image provided by bats uses busybox. This breaks some
tests for two reasons. 'mktemp' template requires at least six X,
otherwise it fails. The 'ln´ error output from busybox differs from the
gnu pendant.

mh182 committed:
Add test case for mock_teardown

Commit:
Document differences to upstream

Commit:
Add Dockerfile to build bats container image including bats-mock

mh182 commented:
If anyone is interested, I maintain a fork of this repo at https://github.com/mh182/bats-mock which includes the changes of this PR.
